### PR TITLE
HW-55005-UI Error Handling (2)

### PR DESCRIPTION
**Pull request by ymoskovko-epam**
- fixed issue of showing the Transfer view when Transfer list is empty 
- Implemented the under-section assertions

### CODE REVIEW

File: Demo/IntegratorAuthenticationProvider.swift [hunk 1]
```diff
@@ -21,12 +21,14 @@ import HyperwalletSDK
 
 public struct IntegratorAuthenticationProvider: HyperwalletAuthenticationTokenProvider {
     private var url: String
-    private let user: String = "userName"
-    private let password: String = "password"
+    private let user: String = "selrestuser@1861681"
+    private let password: String = "Password1!"
+    private let baseUrl = "https://qamaster-hyperwallet.aws.paylution.net"
+    private let userToken = "usr-02d381d3-bf7e-43fe-9d00-02e26b28df57"
     private let session: URLSession
```
Comment (malaw-hw) [inline Demo/IntegratorAuthenticationProvider.swift]: Looks like you forget to remove the test user info

Comment (ymoskovko-epam) [reply Demo/IntegratorAuthenticationProvider.swift]: yes, my fault, fixed

File: Transfer/Sources/View/TransferTableViewFooterView.swift [hunk 2]
```diff
@@ -18,15 +18,15 @@
 
 import UIKit
 
-final class TrasferTableViewFooterView: UITableViewHeaderFooterView {
-    static let reuseIdentifier = "TrasferTableViewFooterViewIdentifier"
+final class TransferTableViewFooterView: UITableViewHeaderFooterView {
+    static let reuseIdentifier = "TransferTableViewFooterViewIdentifier"
```
Comment (jsaini-hw) [inline Transfer/Sources/View/TransferTableViewFooterView.swift]: update TransferTableViewFooterViewIdentifier to transferFooterViewIdentifier

File: Transfer/Sources/View/TransferTableViewFooterView.swift [hunk 3]
```diff
@@ -18,15 +18,15 @@
...
 
     lazy var footerLabel: UILabel = {
         let label = UILabel(frame: .zero)
         label.lineBreakMode = .byWordWrapping
         label.numberOfLines = 0
         label.translatesAutoresizingMaskIntoConstraints = false
-        label.accessibilityIdentifier = "TrasferTableViewFooterLabelIdentifier"
+        label.accessibilityIdentifier = "TransferTableViewFooterLabelIdentifier"
```
Comment (jsaini-hw) [inline Transfer/Sources/View/TransferTableViewFooterView.swift]: update TransferTableViewFooterLabelIdentifier to transferFooterLabelIdentifier

File: Transfer/Sources/View/TransferTableViewFooterView.swift [hunk 4]
```diff
@@ -18,15 +18,15 @@
 
 import UIKit
 
-final class TrasferTableViewFooterView: UITableViewHeaderFooterView {
-    static let reuseIdentifier = "TrasferTableViewFooterViewIdentifier"
+final class TransferTableViewFooterView: UITableViewHeaderFooterView {
```
Comment (jsaini-hw) [inline Transfer/Sources/View/TransferTableViewFooterView.swift]: why do we need custom footerView?

File: Common/Sources/Localization/en.lproj/Localizable.strings [hunk 5]
```diff
@@ -214,6 +214,7 @@
 "transfer_section_header_summary" = "SUMMARY";
 "transfer_button_confirm" = "Confirm";
 "transfer_assert_enter_amount_or_transfer_all" = "Enter amount or select tranfer all funds";
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Common/Sources/Localization/en.lproj/Localizable.strings]: fixed

File: Common/Sources/Localization/en.lproj/Localizable.strings [hunk 6]
```diff
@@ -214,6 +214,7 @@
...
+"transfer_assert_add_a_transfer_method_first" = "Add a transfer method first";
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Common/Sources/Localization/en.lproj/Localizable.strings]: fixed

File: Transfer/Sources/CreateTransferController.swift [hunk 7]
```diff
@@ -69,11 +72,16 @@ public final class CreateTransferController: UITableViewController {
         tableView = UITableView(frame: view.frame, style: .grouped)
         tableView.accessibilityIdentifier = "createTransferTableView"
         tableView.cellLayoutMarginsFollowReadableWidth = false
+        tableView.sectionFooterHeight = UITableView.automaticDimension
+        tableView.estimatedSectionFooterHeight = defaultUIKitFooterHeight
+        tableView.rowHeight = UITableView.automaticDimension
+        tableView.estimatedRowHeight = defaultUIKitRowHeight
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferController.swift]: fixed

File: Transfer/Sources/CreateTransferController.swift [hunk 8]
```diff
@@ -27,6 +27,9 @@ import UIKit
 /// Each transfer will be represented by an auto-generated, non-editable token that can be used
 /// to retrieve the transfer resource.
 public final class CreateTransferController: UITableViewController {
+    let defaultUIKitRowHeight = CGFloat(44.0)
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferController.swift]: removed

File: Transfer/Sources/CreateTransferController.swift [hunk 9]
```diff
@@ -27,6 +27,9 @@ import UIKit
...
+    let defaultUIKitFooterHeight = CGFloat(33.0)
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferController.swift]: removed

File: Transfer/Sources/CreateTransferController.swift [hunk 10]
```diff
@@ -69,11 +72,16 @@ public final class CreateTransferController: UITableViewController {
         tableView = UITableView(frame: view.frame, style: .grouped)
         tableView.accessibilityIdentifier = "createTransferTableView"
         tableView.cellLayoutMarginsFollowReadableWidth = false
+        tableView.sectionFooterHeight = UITableView.automaticDimension
+        tableView.estimatedSectionFooterHeight = defaultUIKitFooterHeight
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferController.swift]: accessibility large text and auto-sized footer

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 11]
```diff
@@ -136,14 +135,14 @@ final class CreateTransferPresenter {
...
-                guard let firstActivatedTransferMetod = result?.data?.first else {
+                guard let firstActivatedTransferMethod = result?.data?.first else {
                     strongSelf.selectedTransferMethod = nil
                     strongSelf.view.hideLoading()
                     strongSelf.view.showCreateTransfer()
                     return
                 }
                 if strongSelf.selectedTransferMethod == nil {
-                    strongSelf.selectedTransferMethod = firstActivatedTransferMetod
+                    strongSelf.selectedTransferMethod = firstActivatedTransferMethod
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: fixed

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 12]
```diff
@@ -136,14 +135,14 @@ final class CreateTransferPresenter {
                 })
 
             case .success(let result):
-                guard let firstActivatedTransferMetod = result?.data?.first else {
+                guard let firstActivatedTransferMethod = result?.data?.first else {
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: removed

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 13]
```diff
@@ -176,19 +175,18 @@ final class CreateTransferPresenter {
                     strongSelf.view.showCreateTransfer()
                 }
             }
+        } else {
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: fixed

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 14]
```diff
@@ -269,6 +267,27 @@ final class CreateTransferPresenter {
         CreateTransferController.FooterSection.allCases.forEach({ view.updateFooter(for: $0) })
     }
 
+    private func areAllFieldsValid() -> Bool {
+        resetErrorMessagesForAllSections()
+        var errors = [HyperwalletError]()
+        // verify empty Amount
+        if amount?.isEmpty ?? true {
+            let error = HyperwalletError(message: "transfer_assert_enter_amount_or_transfer_all".localized(),
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: renamed

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 15]
```diff
@@ -293,6 +312,11 @@ final class CreateTransferPresenter {
                 continue
             }
             switch fieldName {
+            case "destination":
+                if let sectionData = sectionData.first(where: { $0.createTransferSectionHeader == .destination }) {
+                    sectionData.errorMessage = error.message
+                    view.updateFooter(for: .destination)
+                }
             case "amount":
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: updated

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 16]
```diff
@@ -293,6 +312,11 @@ final class CreateTransferPresenter {
                 continue
             }
             switch fieldName {
+            case "destination":
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: obsoleted

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 17]
```diff
@@ -269,6 +267,27 @@ final class CreateTransferPresenter {
         CreateTransferController.FooterSection.allCases.forEach({ view.updateFooter(for: $0) })
     }
 
+    private func areAllFieldsValid() -> Bool {
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: moved

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 18]
```diff
@@ -176,19 +175,18 @@ final class CreateTransferPresenter {
...
 
     // MARK: - Create Transfer Button Tapped
     func createTransfer() {
-        resetErrorMessagesForAllSections()
-        if !transferAllFundsIsOn && amount?.isEmpty ?? true {
-            let error = HyperwalletError(message: "transfer_assert_enter_amount_or_transfer_all".localized(),
-                                         code: "EMPTY_AMOUNT",
-                                         fieldName: "amount")
-            updateFooterContent([error])
+        guard areAllFieldsValid() else {
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: added

File: Transfer/Sources/View/TransferTableViewFooterView.swift (hunk 3 above)
Comment (ymoskovko-epam) [reply Transfer/Sources/View/TransferTableViewFooterView.swift]: updated

File: Transfer/Sources/View/TransferTableViewFooterView.swift (hunk 4 above)
Comment (ymoskovko-epam) [reply Transfer/Sources/View/TransferTableViewFooterView.swift]: https://developer.apple.com/documentation/uikit/uitableviewdatasource/1614994-tableview

The table view uses a fixed font style for section footer titles. If you want a different font style, return a custom view...

also, it is easier to implement auto-resizing, I believe

File: Transfer/Sources/TransferButtonCell.swift [hunk 19]
```diff
@@ -26,27 +26,38 @@ final class TransferButtonCell: UITableViewCell {
     private(set) lazy var buttonLabel: UILabel = {
         let label = UILabel()
         label.translatesAutoresizingMaskIntoConstraints = false
+        label.accessibilityIdentifier = "addTransferNextLabel"
```
[the remark this answers is not in the document]
Comment (ymoskovko-epam) [reply Transfer/Sources/TransferButtonCell.swift]: fixed

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 20]
```diff
@@ -289,23 +282,9 @@ final class CreateTransferPresenter {
...
-                }
-            case "notes":
-                if let sectionData = sectionData.first(where: { $0.createTransferSectionHeader == .notes }) {
-                    sectionData.errorMessage = error.message
-                    view.updateFooter(for: .notes)
-                }
-
-            default:
-                break
+            if let sectionData = sectionData.first(where: { $0.createTransferSectionHeader == .transfer }) {
```
Comment (jsaini-hw) [inline Transfer/Sources/CreateTransferPresenter.swift]: For now let's show all the errors in the footer of transfer section. Even for notes error will be shown under transfer section. Once Willem comes up with concrete requirement then we will create another ticket to handle all those scenarios.

Comment (ymoskovko-epam) [reply Transfer/Sources/CreateTransferPresenter.swift]: Roger that.

**Comment by ymoskovko-epam** [comment]
> Actually the code coverage has gone from 97.46 to 97.41 - @ymoskovko-epam could you kindly fix it

Did you clean the DerivedData folder before? I have the following result:

drawing

**Pull request merged**